### PR TITLE
Task "enable" field

### PR DESCRIPTION
**Pull request by romasku**
Closes https://github.com/neuromation/neuro-flow/issues/50

### CODE REVIEW

File: neuro_flow/expr.py [hunk 1]
```diff
@@ -148,8 +156,44 @@ async def from_json(ctx: CallCtx, arg: str) -> TypeT:
     return cast(TypeT, json.loads(arg))
 
 
+def _check_has_needs(ctx: CallCtx, func_name: str) -> None:
+    try:
+        ctx.root.lookup("needs")
+    except LookupError:
+        raise ValueError(f"{func_name}() is only available inside task definition")
```
Comment (asvetlov) [inline neuro_flow/expr.py:163]:
```suggestion
        raise ValueError(f"{func_name}() is only available inside a task definition")
```

File: neuro_flow/expr.py [hunk 2]
```diff
@@ -148,8 +156,44 @@ async def from_json(ctx: CallCtx, arg: str) -> TypeT:
     return cast(TypeT, json.loads(arg))
 
 
+def _check_has_needs(ctx: CallCtx, func_name: str) -> None:
```
Comment (asvetlov) [inline neuro_flow/expr.py:159]:
```suggestion
def _check_has_needs(ctx: CallCtx, *, func_name: str) -> None:
```

File: neuro_flow/expr.py [hunk 3]
```diff
@@ -148,8 +156,44 @@ async def from_json(ctx: CallCtx, arg: str) -> TypeT:
...
+    for dependency in needs:
+        dep_ctx = needs[dependency]
+        if dep_ctx.result != JobStatus.SUCCEEDED:  # type: ignore
+            return False
+    return True
+
+
+async def failure(ctx: CallCtx) -> bool:
+    _check_has_needs(ctx, func_name="failure")
+    return not success(ctx)
```
Comment (asvetlov) [inline neuro_flow/expr.py:184]: I think `failure()` should return `True` only if dependencies are failed,  for canceled tasks it returns `False`.
The reason is: you can add a code that does something when previous steps are failed but this thing doesn't make sense for canceled workflow.
`always()` exists to handle all statuses.

File: neuro_flow/expr.py [hunk 4]
```diff
@@ -148,8 +156,44 @@ async def from_json(ctx: CallCtx, arg: str) -> TypeT:
...
-    len=alen, nothing=nothing, fmt=fmt, keys=akeys, to_json=to_json, from_json=from_json
+    len=alen,
+    nothing=nothing,
+    fmt=fmt,
+    keys=akeys,
+    to_json=to_json,
+    from_json=from_json,
+    always=always,
+    success=success,
+    failure=failure,
```
Comment (asvetlov) [inline neuro_flow/expr.py:196]: Is there reason for not implementing `cancelled()` function?

File: neuro_flow/expr.py [hunk 5]
```diff
@@ -270,7 +314,8 @@ class Call(Item):
     async def eval(self, root: RootABC) -> TypeT:
         args = [await a.eval(root) for a in self.args]
         try:
-            tmp = await self.func.call(root, *args)  # type: ignore
+            call_ctx = CallCtx(self.start, self.end, root)
```
Comment (asvetlov) [inline neuro_flow/expr.py:317]: Good catch!

File: neuro_flow/expr.py (hunk 4 above)
Comment (romasku) [reply neuro_flow/expr.py:196]: I didn't found any way to cancel started workflow, so at first I think that we do not have concept of cancelled task. But now we do (task with `enable: False`), so thanks for suggestion, I'll add it.

Comment (romasku) [reply neuro_flow/expr.py:208]: I have added `cancelled()`, but I am not sure about it:
I think that it should evaluate to `True` only when the workflow is canceled. I mean, after running something like `neuro-flow cancel batch-id`. I do not think that `task-b` should always in the following workflow:
```
kind: batch
tasks:
  - id: task_a
    enable: false
    image: ubuntu
    preset: cpu-small
    bash: echo "Task A"

  - id: task-b
    needs: [task-a]
    enable: ${{ cancelled() }}
    image: ubuntu
    preset: cpu-small
    bash: echo "Task B"
```
Therefore, we should probably avoid adding `cancelled()` function until the canceling of workflows will be available.

Comment (romasku) [reply neuro_flow/expr.py:208]: Found this issue https://github.com/neuromation/neuro-flow/issues/68. I think that `cancelled()` should raise `NotImplementedError` until that issue is closed.

File: neuro_flow/batch_runner.py [hunk 6]
```diff
@@ -152,15 +155,24 @@ async def process(
                     deps = ctx.graph[tid]
                     needs = {}
                     for dep_id in deps:
-                        dep = finished.get(dep_id)
-                        assert dep is not None
-                        needs[dep_id] = DepCtx(dep.status, dep.outputs)
+                        if dep_id in skipped:
+                            needs[dep_id] = DepCtx(JobStatus.CANCELLED, {})
```
Comment (asvetlov) [inline neuro_flow/batch_runner.py:159]: Not sure if the skipped task should have CANCELLED state

Comment (romasku) [reply neuro_flow/batch_runner.py:159]: I agree, I'll introduce DISABLED task state.

**Review by asvetlov — APPROVED** [review]
LGTM

**Pull request merged**